jax/numpy/lax_numpy.py

```diff
@@ -1148,24 +1148,22 @@ def atleast_3d(*arys):
     return [atleast_3d(arr) for arr in arys]
 
 
-# TODO(mattjj): can this be simplified?
 @_wraps(onp.array)
 def array(object, dtype=None, copy=True, order="K", ndmin=0):
   del copy  # Unused.
   if ndmin != 0 or order != "K":
     raise NotImplementedError("Only implemented for order='K', ndmin=0.")
 
-  if hasattr(object, '__asarray__'):
-    return object.__asarray__(dtype)
-  elif isinstance(object, ndarray):
+  if isinstance(object, ndarray):
     if dtype and _dtype(object) != dtype:
       return lax.convert_element_type(object, dtype)
     else:
       return object
+  elif hasattr(object, '__array__'):
```

Comment:
It's not obvious to me what objects have the `__array__` property but aren't `ndarrays`. Perhaps comment what this is trying to catch?

Reply:
Done. It's for duck typing. This is the same method we override on our DeviceArrays so that numpy knows how to turn them into ndarrays. So this case would fire if one called jax.numpy.array on some class that isn't a subclass of ndarray but implements this method.